### PR TITLE
New Text Mobject

### PR DESCRIPTION
**Pull request by xy-23**
# Motivation

Because the `TextMobject` is the subclass of `TexMobject`,

so it requires some latex-style behaviours in order to add some changes.

(e.g. other languages, font, slant and weight)

So I create `Text` mobject but you can still use `TextMobject`.

# Introduction

`Text` is a subclass of `SVGMobject`.

It uses `cairo` to generate text svgs.

It has a better UTF-8 support, multi-line support, and various options. 

It doesn't support mathematical formula(latex), 

you can use `TextMobject` or `TexMobject`.

You can create a text in manim like:
```
from manimlib.imports import *

class Demo(Scene):
    def construct(self):
        text = Text('Hello, world!')
        self.play(Write(text))
```

introduction

# Features

### text
Unlike `TextMobject`, `Text` only receive one `str`.

When there is only one line, the space(`' '`) in front or back will be ignored.

`'\t'` is not supported.

### color
`text = Text('Hello', color=BLUE)`

color

`color` defined in constants.py or a `str` like `'#FFFFFF'`, default is `WHITE`

### t2c (text2color)
`text = Text('Hello, world!', t2c={'world':BLUE})`

t2c

a `dict` like `{'text':color}` or Accurate mode

### gradient
`text = Text('Hello', gradient=(BLUE, GREEN))`

gradient

a `tuple` like `(color1, color2, color3, ...)`

### t2g (text2gradient)
`text = Text('Hello, world!', t2g={'world':(BLUE, GREEN)})`

t2g

a `dict` like `{'text':gradient}` or Accurate mode

### font
`text = Text('Hello', font='Source Han Sans')`

font

a `str`, the name of font like `'Source Han Sans'`, default is `CONFIG['font']`, defined in constants.py

Be sure you have the font installed!

### t2f (text2font)
`text = Text('Hello, world!', t2f={'world':'Forte'})`

t2f

a `dict` like `{'text':font}` or Accurate mode

Be sure you have the font installed!

### slant
`text = Text('Hello', slant=ITALIC)`

slant

`NORMAL` or `ITALIC`, default is `NORMAL`, defined in constants.py(a `str` actually)

### t2s (text2slant)
`text = Text('Hello, world!', t2s={'world':ITALIC})`

t2s

a `dict` like `{'text':slant}` or Accurate mode

### weight
`text = Text('Hello', weight=BOLD)`

weight

`NORMAL` or `BOLD`, default is `NORMAL`, defined in constants.py(a `str` actually)

other...

### t2w (text2weight)
`text = Text('Hello, world!', t2w={'world':BOLD})`

t2w

a `dict` like `{'text':weight}` or Accurate mode

### size
`text = Text('Hello', size=5)`

size

a number, better larger than 0.1(due to anti-aliasing), irrelevant with MUnit, default is `CONFIG['size']` 

### lsh (line_spacing_height)
`text = Text('Hello\nWorld', lsh=1.5)`

lsh

a number, better larger than 0.1(due to anti-aliasing), irrelevant with MUnit, default is equal with size

### Accurate mode
```
text = Text(
    'Google', 
    t2c={
        '[:1]':'#3174f0', '[1:2]':'#e53125', 
        '[2:3]':'#fbb003', '[3:4]':'#3174f0', 
        '[4:5]':'#269a43', '[5:]':'#e53125',
    }
)
```

accurate_mode

This will help you to choose a specific text just like slicing

### Mobject params:
like fill_color, fill_opacity, stroke_color, stroke_width

## Functions
`text = Text('Hello, world!')`

### set_color (Mobject function)

`text.set_color(BLUE)`

set_color_1

or

`text[7:12].set_color(BLUE)`

set_color_2

### set_color_by_gradient (Mobject function)

`text.set_color_by_gradient(BLUE, GREEN)`

set_color_by_gradient_1

or

`text[7:12].set_color_by_gradient(BLUE, GREEN)`

set_color_by_gradient_2

### set_color_by_t2c
`text.set_color_by_t2c({'world':BLUE})`

set_color_by_t2c

### set_color_by_t2g
`text.set_color_by_t2g({'world':(BLUE, GREEN)})`
set_color_by_t2g

## UTF-8
```
from manimlib.imports import *

script = '''
Hello
你好
こんにちは
안녕하세요
'''

class Demo(Scene):
    def construct(self):
        text = Text(script, font='Source Han Sans')
        self.play(Write(text))
```
utf-8

Be sure you have the font that supports those languages!

### CODE REVIEW

**Comment by xy-23** [comment]
#681

**Comment by Elteoremadebeethoven** [comment]
It works perfectly, the only possible drawback is that the Text to TextMobject transformations do not look correctly, but the transformations between Text to itself and other objects look good. I do add it to my version of Manim, thanks for this!

File: manimlib/constants.py [hunk 1]
```diff
@@ -36,7 +38,8 @@ def initialize_directories(config):
                 "directory were both passed"
             )
 
-    TEX_DIR = config["tex_dir"] or os.path.join(MEDIA_DIR, "Tex")
+    TEX_DIR = config["tex_dir"] or os.path.join(MEDIA_DIR, "texs")
```
Comment (eulertour) [inline manimlib/constants.py]: Please don't change old directories, just create the new one.

File: manimlib/mobject/svg/svg_mobject.py [hunk 2]
```diff
@@ -284,12 +284,33 @@ def handle_transforms(self, element, mobject):
             pass
         # TODO, ...
 
+    def flatten(self, input_list):
+        output_list = []
+        while True:
+            if input_list == []:
+                break
+            for index, i in enumerate(input_list):
```
Comment (eulertour) [inline manimlib/mobject/svg/svg_mobject.py]: This `for` isn't necessary since the if and else blocks both break

**Comment by xy-23** [comment]
@eulertour Updated .

**Comment by kolibril13** [comment]
@xy-23 : can you add your beautiful instruction here:
https://github.com/Elteoremadebeethoven/Manim-community 
??
That would be great, because it is very invisible in a pull request.
It should be easy to find for everyone

**Comment by xy-23** [comment]
@kolibril13 
Thanks, I will, but it has some bugs and improvements need to be finished.
I will do this later until I finish them.

**Comment by Tridu33** [comment]
It works perfectly，Thank you.

**Comment by khengari77** [comment]
is it possible to change the alignment from (left to right) to (right to left ) ?

**Comment by xy-23** [comment]
> is it possible to change the alignment from (left to right) to (right to left ) ?

Sorry, it can't, but you can use multiple Text and align by yourself (manim provides some methods of mobject).
May be I will add this feture later depending on whether the package I used(pycairo) has this function.

**Comment by zamirkhan** [comment]
I really love this addition of the `Text` object, however, I'm seeing some strange artifacts when rendering text in some fonts. Look at the letters b, d, g, o, p, q below with Roboto Condensed font.

Screen Shot 2019-11-11 at 11 07 32 PM

Also see the B and P below, with Noto Sans. These are just examples of what I see and not the only artifacts I encounter. 

Screen Shot 2019-11-11 at 11 13 39 PM


Any idea what I might be doing wrong?

**Comment by Elteoremadebeethoven** [comment]
Try `Text(”text”,stroke_width=0)`

**Comment by zamirkhan** [comment]
That appears to have worked. Thank you!

**Comment by qo4on** [comment]
Doesn't work
image

**Comment by ahmedafifkhan** [comment]
**How can I write different size in one string?**

For example: I want "Hello World how are you?" to be looking like
*"Hello World"* size=2 and *"how are you?"* size=1

**Comment by Tony031218** [comment]
@ahmedafifkhan you can scale "how are you" manually.
```python
text = Text("Hello World how are you?", size=2)
text[12:].scale(0.5, about_edge=LEFT)
```

**Comment by codecxx** [comment]
good

**Comment by Galaxyjia** [comment]
cool this solve my problem

**Comment by OsmanElhassan** [comment]
hi i'm trying to render in Arabic, i followed your tutorial and it's really helpful. l just got one problem as it shows in the picture, the word is reversed and disconnected. i would be grateful if you can help me out. thanks in advance  @xy-23 
Demo

**Pull request merged**